src/server/pfs/server/driver.go

```diff
@@ -2588,6 +2632,122 @@ func (d *driver) getFile(pachClient *client.APIClient, file *pfs.File, offset in
 	return grpcutil.NewStreamingBytesReader(getBlocksClient, nil), nil
 }
 
+// getFiles takes the objReader byte stream of object storage references generated in
+// getFile and parses it into GetFileResponses that are then acted upon individually
+// by the callback function f.
+func (d *driver) getFiles(pachClient *client.APIClient, objReader io.Reader, file *pfs.File, f func(*pfs.GetFileResponse) error) (retErr error) {
+	commitInfo, err := d.inspectCommit(pachClient, file.Commit, pfs.CommitState_STARTED)
+	if err != nil {
+		return err
+	}
+	var buf bytes.Buffer
+	maxData := int64(grpcutil.MaxMsgSize / 10)
+
+	// parseFiles generates one or more GetFileResponse for the specific node
+	// at the provided path p. The first GetFileResponse for a given node, contains up to
+	// the first 2MB of data as well as FileInfo metadata, while the rest only contain
+	// (up to 2MB of) data and nil FileInfo.
+	// The fileInfo argument, if provided (i.e.input repo that could potentially have header/footer),
+	// is used as the GetFileResponse for the first block.
+	parseFiles := func(p string, node *hashtree.NodeProto, fileInfo *pfs.FileInfo) error {
+		if node.FileNode == nil {
+			return nil
+		}
+
+		if fileInfo == nil {
+			fileInfo = nodeToFileInfo(commitInfo, p, node, true)
+		}
+
+		remainingData := int64(fileInfo.SizeBytes)
+		i := 0
+		for remainingData > 0 || i == 0 {
+			var err error
+			if remainingData <= maxData {
+				_, err = io.CopyN(&buf, objReader, remainingData)
+			} else {
+				_, err = io.CopyN(&buf, objReader, maxData)
+			}
+			if err != nil && err != io.EOF {
+				return err
+			}
+			gfr := &pfs.GetFileResponse{
+				Value:    buf.Bytes(),
+				FileInfo: fileInfo,
+			}
+			fileInfo = nil
+			if err := f(gfr); err != nil {
+				return err
+			}
+			remainingData -= maxData
+			buf.Reset()
+			i++
+		}
+		return nil
+	}
+
+	// Handle commits to input repos
+	if commitInfo.Provenance == nil {
+		tree, err := d.getTreeForFile(pachClient, client.NewFile(file.Commit.Repo.Name, file.Commit.ID, ""))
+		if err != nil {
+			return err
+		}
+		err = tree.Glob(file.Path, func(p string, node *hashtree.NodeProto) error {
+			if node.DirNode != nil {
+				return tree.Walk(p, func(pf string, nf *hashtree.NodeProto) error {
```

Comment:
Shouldn't there be a check for the recursive flag here? I think you could be in a situation where if it was not set, you could parse the files incorrectly.

Reply:
Done.